File: code/design-patterns/structural/proxy/go/proxy.go
package main

import (
	"fmt"
)

// ServiceInterface represents the interface for the service
type ServiceInterface interface {
	handleRequest(url string) string
}

// WebServer represents the actual service
type WebServer struct{}

// handleRequest handles the request directly in the WebServer
func (ws *WebServer) handleRequest(url string) string {
	return fmt.Sprintf("WebServer handling request for: %s", url)
}

// Nginx represents the proxy for the WebServer
type Nginx struct {
	webServer   *WebServer
	rateLimiter map[string]int
	cache       map[string]string
	maxRequests int
}

// NewNginx creates a new Nginx proxy with a given maxRequests limit
func NewNginx(maxRequests int) *Nginx {
	return &Nginx{
		webServer:   &WebServer{},
		rateLimiter: make(map[string]int),
		cache:       make(map[string]string),
		maxRequests: maxRequests,
	}
}

// handleRequest controls access, rate limits, and caches requests
func (nginx *Nginx) handleRequest(url string) string {
	if !nginx.controlAccess(url) {
		return "Access denied"
	}

	if response, found := nginx.cache[url]; found {
		fmt.Println(response)
		return fmt.Sprintf("Returning cached response for: %s", url)
	}

	response := nginx.webServer.handleRequest(url)
	nginx.cacheRequest(url, response)
	return response
}

// controlAccess checks the rate limit for the given url
func (nginx *Nginx) controlAccess(url string) bool {
	if count, exists := nginx.rateLimiter[url]; exists {
		if count >= nginx.maxRequests {
			return false
		}
		nginx.rateLimiter[url]++
	} else {
		nginx.rateLimiter[url] = 1
	}
	return true
}

// cacheRequest caches the response for the given url
func (nginx *Nginx) cacheRequest(url, response string) {
	nginx.cache[url] = response
}

File: code/design-patterns/structural/proxy/python/proxy.py
from typing import Dict


class ServiceInterface:
    def handle_request(self, url: str) -> str:
        pass


class WebServer(ServiceInterface):
    def handle_request(self, url: str) -> str:
        return f"WebServer handling request for: {url}"


class Nginx(ServiceInterface):
    def __init__(self, max_requests: int):
        self.web_server = WebServer()
        self.rate_limiter: Dict[str, int] = {}
        self.cache: Dict[str, str] = {}
        self.max_requests = max_requests

    def handle_request(self, url: str) -> str:
        # Check if rate limit exceeded
        if not self.control_access(url):
            return "Access denied"
        # Check if in cache
        if url in self.cache:
            return f"Returning cached response for: {url}"

        response = self.web_server.handle_request(url)
        self.cache_request(url, response)
        return response

    def control_access(self, url: str) -> bool:
        if url in self.rate_limiter:
            if self.rate_limiter[url] >= self.max_requests:
                return False
            self.rate_limiter[url] += 1
        else:
            self.rate_limiter[url] = 1
        return True

    def cache_request(self, url: str, response: str):
        self.cache[url] = response
